Sources/UserScript/UserScriptMessaging.swift

```diff
@@ -0,0 +1,263 @@
+        case error(message: String)
+    }
+
+    /// Convert incoming messages, into an Action.
+    ///
+    /// Conditions for `error`:
+    ///  - does not contain `featureName`, `context` or `method`
+    ///  - delegate not found for `featureName`
+    ///  - origin not supported, due to a feature's configuration
+    ///  - delegate failed to provide a handler
+    ///
+    /// Conditions for `respond`
+    ///  - no errors
+    ///  - contains an `id`
+    ///
+    /// Conditions for `notify`
+    ///  - no errors
+    ///  - does NOT contain an `id`
+    public func messageHandlerFor(_ message: WKScriptMessage) -> Action {
+
+        /// first, check that the incoming message is roughly in the correct shape
+        guard let dict = message.messageBody as? [String: Any],
+              let featureName = dict["featureName"] as? String,
+              let context = dict["context"] as? String,
+              let method = dict["method"] as? String
+        else {
+            return .error(message: "The incoming message was not valid - one or more of 'featureName', 'name'  or 'context' was missing")
+        }
+
+        /// Now try to match the message to a registered delegate
+        guard let delegate = callbacks[featureName] else {
+            return .error(message: "feature named `\(featureName)` was not found")
+        }
+
+        /// Check if the selected delegate accepts messages from this origin
+        guard delegate.allowedOrigins.isAllowed(hostProvider.hostForMessage(message)) else {
+            return .error(message: "invalid origin")
+        }
+
+        /// Now ask the delegate to provide the handler
+        guard let handler = delegate.handlerFor(method) else {
+            return .error(message: "the incoming message is ignored because the feature `\(featureName)` couldn't provide a handler for method `\(method)`")
+        }
+
+        /// just send empty params if absent
+        var methodParams: Any = [:] as Any
+        if let params = dict["params"] {
+            methodParams = params
+        }
+
+        /// if the incoming message had an 'id' field, it requires a response
+        if let id = dict["id"] as? String {
+            let incoming = RequestMessage(context: context, featureName: featureName, id: id, method: method, params: methodParams)
+            return .respond(handler: handler, request: incoming)
+        }
+
+        /// If we get this far, we are confident the message was in the correct format
+        /// but we don't think it requires a response. Therefor we treat it as a fire-and-forget notification
+        let notification = NotificationMessage(context: context, featureName: featureName, method: method, params: methodParams)
+        return .notify(handler: handler, notification: notification)
+    }
+
+    /// Perform the side-effect described in an action
+    public func execute(action: Action, original: WKScriptMessage, completion: @escaping (String) -> Void) throws {
```

Comment:
I'm sorry for flipping the table but we established we should no longer use the old api for handling async code but rather async/await - I'll be glad to assist you on the changes

Reply:
No that's great, it will be much better as async/await 👌🏻